### DISCUSSION TITLE
Lost connection to server during query

### DISCUSSION THREAD

**Discussion by Quentmgt** [sqlalchemy/sqlalchemy#10639] [Usage Questions] [answered]
Hello,

I have curentely an app with sqlalchemy 1.4.41 and a mariadb server 11.2.1 on Windows
For testing i have configured the 'interactive_timeout'  and 'wait_timeout' at 60 on my mariadb server.

I use this to connect to my server,
`engine = create_engine(f"mariadb+pymysql://user:pass@127.0.0.1/db01", echo=True, pool_recycle=30, pool_pre_ping=True)`
and i have tried with mysql driver:
`engine = create_engine(f"mysql://user:pass@127.0.0.1/db01", echo=True, pool_recycle=30, pool_pre_ping=True)`

and after 60 seconds, if i make select request for example, i have this error:

> sqlalchemy.exc.OperationalError: (pymysql.err.OperationalError) (2013, 'Lost connection to MySQL server during query ([WinError 10053] An established connection was aborted by the software in your host machine)')

I don't know if it's a bug or me not understanding why it don't make a new connection after 30 seconds of innactivity.

Thanks in advance for your help

**Comment by CaselIT** [comment]
hi,

without any example it's not easy to reply. you seem to be keeping a single connection checked out for more than 60 seconds. the pre-print and recycle options only take effect when checking out connections from the pool, (so when doing engine.connect())

**Comment by Quentmgt** [comment]
```python
engine = create_engine(f"mariadb+pymysql://user:pass@127.0.0.1/db01", echo=True, pool_recycle=30, pool_pre_ping=True)
Session = sessionmaker(bind=engine)
session = Session()
Base = declarative_base()


class UserData(Base):
    __tablename__ = "user_data"
    id = Column(Integer, primary_key=True, default=get_new_id())
    user_id = Column(Integer)
    user_name = Column(String(30))
    ...

    @staticmethod
    def get_user(tluser_id,) -> UserData:
        user = session.query(UserData).filter(UserData.user_id == tluser_id).first()
        return user
```
Is the example enough clear ?
Here, the `get_user` method is called by an externatl telegram event.
and when i trigger the execution of this method after 60 seconds of innactivity, the exception appear

My connection remains open all the time, it is an mistake ? it is not more efficient to do this ?

**Reply by CaselIT** [reply]
well you have a module level session there: `session = Session()`.
That will keep the transaction open until you close it, so that's why the connection expires after a bit

**Reply by CaselIT** [reply, marked as answer]
you may also want to look at https://docs.sqlalchemy.org/en/20/orm/session_basics.html#session-faq-whentocreate

**Reply by Quentmgt** [reply]
Okay, now i see why you asked an example.

For all my database objects, i have an 'db' module 

Here the '\_\_init\_\_' of this module 
```python
from sqlalchemy import create_engine, MetaData
from sqlalchemy.orm import sessionmaker, declarative_base

engine = create_engine(f"mariadb+pymysql://user:pass@127.0.0.1/db01", echo=True, pool_recycle=30, pool_pre_ping=True)

Session = sessionmaker(bind=engine)
Metadata = MetaData()
session = Session()

Base = declarative_base()

from db.UserData import UserData
from db.UserSubscription import UserSubscription
...
```

And the UserData file
```python
from __future__ import annotations
from sqlalchemy import Column, Integer, String, func, Table
from db import Base, session

class UserData(Base):
    __tablename__ = "user_data"
    id = Column(Integer, primary_key=True, default=get_new_id())
    user_id = Column(Integer)
    user_name = Column(String(30))
    ...

    @staticmethod
    def get_user(tluser_id) -> UserData:
        user = session.query(UserData).filter(UserData.user_id == tluser_id).first()
        return user
```

When i want to use the database in program, i import the module and the session stay open. It allows me to query directely with `UserData.get_user()`
```python
from db import *
...

def main_menu(tluser_id):
    current_user = UserData.get_user(tluser_id)
    
    message = f"Hello {current_user.user_name}! Your id is {current_user.id}."
    return message
```

If i have corectely understand the doc, an better way to do would be to remove the `session = Session()` of my module init file, then correct my UserData class to have in parameters the session
```python
    @staticmethod
    def get_user(db_session, tluser_id) -> UserData:
        user = db_session.query(UserData).filter(UserData.user_id == tluser_id).first()
        return user
```
and in my program, at method call, create the session, make operations i need and close the session
```python
def main_menu(tluser_id):
    db_session = Session()
    current_user = UserData.get_user(db_session, tluser_id)
    
    message = f"Hello {current_user.user_name}! Your id is {current_user.id}."
    # Some others database operations....
    db_session .close()
    return message
```

Is it a good way to do ?

Thanks for your help 🙏

**Reply by CaselIT** [reply]
In general yes, this way you can better control the interaction with the db starts and where it ends

**Reply by Quentmgt** [reply]
Ok I think I understand now, thank you very much !
